### DISCUSSION TITLE
How to createMany with a relation?  Looping CREATE and getting  "too many connections"?

### DISCUSSION THREAD

**Discussion by fotoflo** [prisma/prisma#17472] [Q&A] [answered]
hello!  Let me see if i got this straight - I have what i believe to be a very simple issue:

1. I create a batch of reports scores
2. I create a bunch of scores in the batch
    - Because you can't use createMany with "nested relations"... i have to loop through this simple call
4. I run out of connections all the time

Simplified Schema
```
model reportingBatch {
  id                    BigInt                  @id @unique(map: "id") @default(autoincrement()) @db.UnsignedBigInt
  status                ReportingStatus
  createdAt             DateTime                @default(now()) @db.Timestamp(0)
  scores                scores[]
}

model scores {
  id                BigInt          @id @unique(map: "id") @default(autoincrement()) @db.UnsignedBigInt
  reportingBatch    reportingBatch  @relation(fields: [reportingBatchId], references: [id], onDelete: Cascade)
  reportingBatchId  BigInt          @db.UnsignedBigInt
  title             String?         @db.VarChar(2048)
  score             Int?
  createdAt             DateTime                @default(now()) @db.Timestamp(0)
}
```

Now i can create a score like this:
```typescript
async function createScores(batch, data){
    return await prisma.scores.create({
      data: {
        reportingBatch: {
          connect: {
            id: batch.id
          },
        },
        ...data,
      },
    });

}
```

But i can't use createMany because of the connect call (no create many with "nested relations" - isn't this a simple relation?) so then i have to throw the above in a loop:

```typescript
async function createScoresFromArray(batch, data){
 const promises = data.forEach( item => {
      return await createScores(batch, item)
   });
 return await Promise.all(promises)
}
```

 Which, beside being sub-optimally performant, runs out of connections before my test suite finishes running... is there a better way to do it? How do i avoid running out of connections? Seems like a simple insert many, with the batchId in the table... am i missing something?

**Comment by nurul3101** [comment, marked as answer]
Hello @fotoflo 👋 

If you just need to connect an id, can you try directly setting the foreign key if you are sure that it exists?
Here's an example of what I am suggesting:
```typescript
import { Prisma, PrismaClient } from "@prisma/client";

const prisma = new PrismaClient();

// A `main` function so that you can use async/await
async function main() {
  await prisma.reportingBatch.create({
    data: {
      id: 1,
    },
  });

  await prisma.scores.createMany({
    data: [
      {
        reportingBatchId: 1,
        score: 1,
        title: "title1",
      },
    ],
  });
}

main()
  .catch((e) => {
    throw e;
  })
  .finally(async () => {
    await prisma.$disconnect();
  });
```
Based on your schema, first a reporting batch is created and then a createMany query on scores model directly sets the foreign key `reportingBatchId` instead of using connect. Would this work for your case?

**Reply by fotoflo** [reply]
Yesss! thank you @nurul3101, it appears to work Thank you! I think i tried that before doing connect and it didn't work for whatever reason lol!

So then my question is what is the point of connect?

**Reply by nurul3101** [reply]
You could directly set the foreign key or use `connect`. Both do the same thing and should work.
However, it is better to use connect to link a foreign key because prisma would throw an error in case it cannot find the connecting record, so connect would provide an extra layer of validation.

**Comment by juanramoncarceles** [comment]
In my case I have a many-to-many relation, so I don't have an explicit foreing key (also I'm using Planet Scale so no foreign keys anyway) and I was using `connect` with an array value. So I guess the only way for me is to not use `createMany` and instead loop and call `create` ? 😔

```
model Collaborator {
  id                String         @id
  name              String
  email             String?
  categories        Category[]
}
```

```js
db.collaborator.createMany({
    data: collaborator.map((col) => ({
        id: ID.Collaborator(),
        name: lead.name,
        email: lead.email,
        // ERROR HERE
        // Invalid `prisma.collaborator.createMany()` invocation: ...
        // Unknown argument `categories`. Available options are marked with ?
        categories: {
            connect: lead.categories.map((id) => ({ id })), 
        },
    })),
});
```

**Reply by carlonoelle** [reply]
Same problem +1

**Reply by nqnhan1403** [reply]
The same with me +1

**Reply by BladimirBasilio** [reply]
+1

**Reply by Etep15** [reply]
+1

**Reply by DASMACHETE** [reply]
+1

**Reply by usmanchlorine** [reply]
+1

**Reply by iypavlov** [reply]
+1

**Reply by FirasLatrech** [reply]
+1

**Reply by CauaMatheus** [reply]
+1

**Reply by turtletongue** [reply]
+1

**Reply by aidan-rypens** [reply]
+1

**Reply by mvp-studio-repo** [reply]
just faced this issue, this should be raised as it can be critical in case of testing or seeding, could you put on a eye on it @janpio ?

**Reply by janpio** [reply]
Then please raise it as a bug issue, optimally with a reliable reproduction so we can experience the problem ourselves. Thanks.
(Make sure to fill the Github issue template so we have all the information needed)

**Reply by muchisx** [reply]
does this already have an open issue/feature request I can open one if needed, just don't want to make a duplicate

**Reply by mvp-studio-repo** [reply]
@muchisx this issue seem to not be duplicated, if you feel like opening the bug issue with the exemple please do ! We will thanks you much :)

**Reply by divyaswormakai** [reply]
+1

**Reply by CodrinSocol** [reply]
+1

**Reply by birgersp** [reply]
+100

**Reply by mmartinez-hd** [reply]
+1

**Reply by Qbason** [reply]
+1

**Reply by fzn0x** [reply]
@janpio Hi, do you have any plans to include this feature? I think this is critical, createMany can't create relation with it.

**Reply by janpio** [reply]
Yes, there is a feature request over at https://github.com/prisma/prisma/issues/5455

**Reply by sovetski** [reply]
same here with the latest version of Prisma

**Reply by Tony1023** [reply]
+1

**Comment by antoine13330** [comment]
Hi guys, as u can see at the bottom of this section [https://www.prisma.io/docs/orm/prisma-client/queries/transactions#transaction-api](https://www.prisma.io/docs/orm/prisma-client/queries/transactions#transaction-api) `$transaction` is implicitly the solution given by the Prisma team 

> As Prisma Client evolves, use cases for the $transaction([]) API will increasingly be replaced by more specialized bulk operations (such as createMany) and nested writes.

So I came with this solution and tested it with 2_000 items and was blazing fast surprisingly : 

```typescript
await prisma.$transaction(endpoints.map((endpoint) => {
          return prisma.apiEndpoint.create({
              data: {
                  ...endpoint,
                  body : endpoint.body as JsonObject,
                  responses: {
                      createMany: {
                          data: endpoint.responses
                     }
                 }
            }
        })
}));
```

Hope this helped someone bypassing this problem ? But prisma will I hope soon give us this feature. They know it's lacking in their product as they said.

**Reply by divyaswormakai** [reply]
We opted out of using a crm for our usecase as a result but this looks like very viable solution. Will refer to this in the future. Thanks a bunch.

**Reply by virasasan** [reply]
In the case when the "owning" objects and all its "owned" objects are being created in the same operations, the suggested solutions would require them to be created in different database transactions. And as far as I can tell, using $transaction won't work in this case, because you first need to create the owning object, get its id, and use it in creating the "owned" objects.

So my question is: is there a way to create the "owning" object and its many "owned" objects all in one database transaction?
